src/core/qgsmaplayer.h

```diff
@@ -882,6 +882,25 @@ class CORE_EXPORT QgsMapLayer : public QObject
      */
     bool isRefreshOnNotifyEnabled() const { return mIsRefreshOnNofifyEnabled; }
 
+///@cond PRIVATE
+#ifndef SIP_RUN
+
+    /**
+     * Set last rendering time.
+     *
+     * \note not available in Python bindings
+     */
+    void setLastRenderingTime( double time ) { mLastRenderingTime = time; }
+
+    /**
+     * Get last rendering time.
+     *
+     * \note not available in Python bindings
+     */
+    double lastRenderingTime() const { return mLastRenderingTime; }
+#endif
+///@endcond
+
```

Comment:
I am wondering if this is a good approach to assign last rendering time information to a map layer. There can be multiple 2D map canvases active, there could be map overview widget, there could be 3D map canvas that renders layers into tiles - all of them using possibly quite different configurations, running at the same time in background, so there may be conflicting updates. The question is whether we really need to worry about that...

Reply:
I wondered the same... but I think it's harmless at worst. Let's give this approach a test run and see how it works in practice before adding extra complications like reporting render times back to the canvas.

Reply:
I thought there was a maplayer object per layer and per canvas ? If not, then perhaps a QMap< QgsMapCanvas*, int > would do

> The question is whether we really need to worry about that..

As I mentioned in my email, I've a particular use case where job previewing clearly decreases performance.

Reply:
> I thought there was a maplayer object per layer and per canvas ? If not, then perhaps a QMap< QgsMapCanvas*, int > would do

No, map layers are per project and can be shared between canvases/composers/etc. The QgsMapCanvas based map won't work either here because canvas is in gui, and map layer in core.

This could potentially be returned as an alternative to QgsMapRendererJob::renderingTime, which would instead return a per-layer render time. The canvas could then collect this and use it to determine which layers to render in its next preview jobs...

Reply:
Might be reasonable to include a flag in the render job if it's for a preview job. Then explicitly these jobs can be supervised.

Reply:
> Might be reasonable to include a flag in the render job if it's for a preview job. Then explicitly these jobs can be supervised.

That's already there - there's a flag in QgsRenderContext which indicates preview jobs (with the intention that rendering can take "short cuts" if available)